src/snowflake/snowpark/lineage.py

```diff
@@ -0,0 +1,454 @@
+    UPSTREAM = "upstream"
+    BOTH = "both"
+
+    def __str__(self):
+        return self.value
+
+
+class _EdgeType(Enum):
+    """
+    Types of edges for lineage tracing.
+    """
+
+    DATA_LINEAGE = "DATA_LINEAGE"
+    OBJECT_DEPENDENCY = "OBJECT_DEPENDENCY"
+
+    @classmethod
+    def list(cls):
+        return [member.value for member in cls]
+
+
+class _ObjectField:
+    """
+    Defines static fields used to reference object properties in DGQL query and response.
+    """
+
+    DOMAIN = "domain"
+    REFINED_DOMAIN = "refinedDomain"
+    USER_DOMAIN = "userDomain"
+    NAME = "name"
+    PROPERTIES = "properties"
+    SCHEMA = "schema"
+    DB = "db"
+    STATUS = "status"
+    CREATED_ON = "createdOn"
+    PARENT_NAME = "ParentName"
+    VERSION = "version"
+
+    # A list of fileds queried on each object in the lineage.
+    GRAPH_ENTITY_PROPERTIES = [
+        DOMAIN,
+        REFINED_DOMAIN,
+        USER_DOMAIN,
+        NAME,
+        PROPERTIES,
+        SCHEMA,
+        DB,
+        STATUS,
+        CREATED_ON,
+    ]
+
+
+class _DGQLFields:
+    """
+    Contains static definitions of field names used in DGQL queries and responses.
+    """
+
+    DATA = "data"
+    NODE = "V"
+    EDGE = "E"
+    SOURCE = "S"
+    TARGET = "T"
+    OUT = "OUT"
+    IN = "IN"
+
+
+class _UserDomain:
+    """
+    Domains in the user context that logically maps to different snowflake objects.
+    """
+
+    FEATURE_VIEW = "FEATURE_VIEW"
+    MODEL = "MODULE"
+
+
+class _SnowflakeDomain:
+    """
+    Snowflake object domains relevant for querying lineage.
+    Note: This is a subset and does not include all possible domains.
+    """
+
+    TABLE = "TABLE"
+    MODULE = "MODULE"
+    DATASET = "DATASET"
+
+
+class Lineage:
+    """
+    Provides methods for exploring lineage of Snowflake objects.
+    To access an object of this class, use :attr:`Session.lineage`.
+    """
+
+    def __init__(self, session: "snowflake.snowpark.session.Session") -> None:
+        self._session = session
+        self.user_to_system_domain_map = {
+            _UserDomain.FEATURE_VIEW: _SnowflakeDomain.TABLE,
+            _UserDomain.MODEL: _SnowflakeDomain.MODULE,
+        }
+        self.versioned_object_domains = {
+            _UserDomain.FEATURE_VIEW,
+            _UserDomain.MODEL,
+            _SnowflakeDomain.DATASET,
+        }
+
+    def _build_graphql_query(
+        self,
+        object_name: str,
+        object_domain: str,
+        edge_directions: list,
+        object_version: Optional[str] = None,
+    ) -> str:
+        """
+        Constructs a GraphQL query for lineage tracing based on the specified parameters.
+        """
+        properties_string = ", ".join(_ObjectField.GRAPH_ENTITY_PROPERTIES)
+        edge_types_formatted = ", ".join(_EdgeType.list())
+
+        parts = []
+        edge_template = "{direction}: {edge_key}(edgeType:[{edge_types}],direction:{dir}){{{source_key} {{{properties}}}, {target_key} {{{properties}}}}}"
```

Comment:
put it as a class constant?

Reply:
We dont use this in any other function right now. Its only useful in query generation function